### PR TITLE
.github: Add pull request template

### LINKED ISSUE #7538: Pull request checklist
**Is your feature request related to a problem? Please describe.**

Checklists help with guiding a process to completion. They ensure quality and make their users confident, because they won't forget certain aspects of the process.

**Describe the solution you'd like**

Add a pull requests checklist to the pull request template and/or "maintainer handbook" (the manual or `maintainers/*.md`)

**Describe alternatives you've considered**

 - Keep forgetting to do certain things leading to
   - bloat
   - regressions
   - forgotten documentation
   - uncertainty as to why changes to the code were made
   - prs getting slowed by delays, reminders and more delays

Have a github action that posts the checklist to each PR. This is a bit more robust as contributors may mess with the checklist.

**Additional context**

 - I forgot to ask for documentation in #7314.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by roberth**
Closes #7538

### CODE REVIEW

**Comment by blaggacao** [comment]
Maybe relevant. Maybe o ky intended for _issues': https://github.com/NixOS/nix/pull/7387

**Comment by fricklerhandwerk** [comment]
I strongly welcome to introduce a checklist.

My gut reaction to putting it into the PR description is aversion though, and I'll try to pick it apart here.

Some considerations and possible alternatives:
- like in Nixpkgs, use labels to express required tasks and PR state
  - this has some potential drawbacks, such as information on what the labels mean only being visible somewhere else
- including a checklist in the PR description introduces noise for PR authors
  - maybe indeed adding an action that posts it immediately as a comment is better
- the checklist could be in the maintainer/contributor guide and we just work it off as a routine
  - this has a similar disadvantage of living in a different place than where the action happens, but would be close to other information for contributors
- a combination of the above, such as a checklist in the manual, a team routine to add labels during review, and a (set of) action(s) to act on the labels in order to produce notifications that carry more context
  - for example, if maintainers add a label "needs release notes", after some delay the bot would post a friendly comment addressed to the author to add release notes, including a link to the checklist in the manual. when the branch is updated to change release notes, that label is replaced by "has release notes"

**Comment by roberth** [comment]
I agree that it's not optimal, but I do think it's an improvement. It's a simple solution, and I am not eager to experiment with labels and actions.
A manual process is the status quo and it's too easy to use it improperly or forget it completely.

> including a checklist in the PR description introduces noise for PR authors

It is also a reminder of what the criteria for a good pr are. Surely not all, but at least some contributors will try to comply with the checkboxes when they see them. That helps everyone achieve their goals.

> after some delay the bot would post a friendly comment

Any delay will make the contributor have to revisit the PR. Often the change is easy to make when you're still working on it, but it takes energy to revisit it and switch contexts.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+##### Checklist for maintainers
+
+<!-- Contributors: please leave this as is -->
+<!-- Maintainers: tick if completed or explain if not relevant -->
+
+ - [ ] is the idea good? has it been discussed by the Nix team?
+ - [ ] unit tests
+ - [ ] functional tests (`tests/**.sh`)
+ - [ ] documentation in the manual
+ - [ ] documentation in the code (if necessary; ideally code is already clear)
```
Comment (fricklerhandwerk) [inline .github/PULL_REQUEST_TEMPLATE.md:20]:
```suggestion
- [ ] code and comments are self-explanatory
```

I was trying to make it more tangible what "documentation in the code" is supposed to mean, but it's really hard to come up with something clear-cut.

I want code to be "readable" in a sense that if you read it, it answers all the questions about itself without having to dig around. That doesn't necessarily mean there need to be lots of comments. The symbol names should allow one to easily infer what they mean, comments should explain why the code is written the way it is, or explain what it does if there is no clean way of expressing it in the code itself. There is plenty of material out there that explains really well why these (and others) are qualities of readable code, and eventually we should collect these resources for reference.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+##### Checklist for maintainers
+
+<!-- Contributors: please leave this as is -->
+<!-- Maintainers: tick if completed or explain if not relevant -->
+
+ - [ ] is the idea good? has it been discussed by the Nix team?
+ - [ ] unit tests
+ - [ ] functional tests (`tests/**.sh`)
+ - [ ] documentation in the manual
```
Comment (fricklerhandwerk) [inline .github/PULL_REQUEST_TEMPLATE.md:19]:
```suggestion
 - [ ] user-facing documentation is appropriate
```

We may want to be more precise about what "appropriate" means, maybe there is a better word...

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+
+
+
+##### Checklist for maintainers
+
+<!-- Contributors: please leave this as is -->
+<!-- Maintainers: tick if completed or explain if not relevant -->
+
+ - [ ] is the idea good? has it been discussed by the Nix team?
```
Comment (fricklerhandwerk) [inline .github/PULL_REQUEST_TEMPLATE.md:16]:
```suggestion
 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
```

To more closely match the [team handbook](https://github.com/NixOS/nix/tree/eaa20f25747eb025c7c4e69fa83f0e455a65393f/maintainers#project-board-protocol).

I thought about writing "approved", but that sounds a lot like bureaucracy.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+<!-- An issue reference or explanation why the change is desirable. -->
+<!-- This speeds up step 1 of the process. -->
+
+
+
+
+
+
+##### Checklist for maintainers
```
Comment (fricklerhandwerk) [inline .github/PULL_REQUEST_TEMPLATE.md:11]: I think we should be more precise about what we want to have here.

```suggestion
# Motivation
<!-- Briefly explain what the change is about and why it is desirable. -->

# Context
<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Checklist for maintainers
```

While it's probably undesirable to have large headings in the description, it also doesn't really hurt to make the sections more distinct. After all, it's just a presentational aspect.

**Review by fricklerhandwerk — COMMENTED** [review]
@roberth agreed, having something is better than nothing. Made some suggestions how to make it more precise and hopefully more useful.

Maybe in the long run we should provide links to materials that go more into detail what a good contribution looks like and why that's so. One that comes to mind, because I was involved in reviewing it is [Getting Things Merged](https://www.tweag.io/blog/2022-04-14-getting-things-merged/#the-technical-aspect-of-getting-things-merged) by @smelc, but there are surely more, other high quality texts on different aspects of the process.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (blaggacao) [reply .github/PULL_REQUEST_TEMPLATE.md:19]: It's a thing of balance, neither too verbose, nor too terse. It's hard to capture where to strike that balance without really having a clearly defined "user/reader persona" for which one is supposed to optimize.

It's a problem, indeed. "balanced" suffers from the same imprecision, but better captures that it's a judgemental trade off.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 3 above)
Comment (blaggacao) [reply .github/PULL_REQUEST_TEMPLATE.md:16]: `agreed on idea` &mdash;  `agreed on idea & design` I think the word "design" adds value around here.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (blaggacao) [reply .github/PULL_REQUEST_TEMPLATE.md:11]: It may even be worthy to dedicate a whole (potentially blankable) section explicitly to design. It nudges people to possibly at least quickly revise their drawing board when they started to hack out  of an itch or momentum and feel really enthusiastic about their adrenaline rush.

```
# Design

<!-- Explain the implementation route and design that you've chosen and why. -->
```

Comment (fricklerhandwerk) [reply .github/PULL_REQUEST_TEMPLATE.md:11]: That was my first idea, too, but that requires additional work for trivial PRs, which I think we'd want to have many of and which should require a trivial amount of effort. If the PR is complex anyway, authors may as well be inclined to make a proper write-up.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (roberth) [reply .github/PULL_REQUEST_TEMPLATE.md:22]: I don't understand why this is an improvement.
It seems a bit wordy and we already have "tick if completed or explain if not relevant".

Comment (fricklerhandwerk) [reply .github/PULL_REQUEST_TEMPLATE.md:22]: With the additional information in the other notes its indeed good as it is.

**Review by fricklerhandwerk — APPROVED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+<!-- Non-trivial change: Briefly outline the implementation strategy. -->
+
+<!-- Invasive change: Discuss alternative designs or approaches you considered. -->
+
+<!-- Large change: Provide instructions to reviewers how to read the diff. -->
+
+# Checklist for maintainers
+
+<!-- Contributors: please leave this as is -->
+<!-- Maintainers: tick if completed or explain if not relevant -->
```
Comment (roberth) [inline .github/PULL_REQUEST_TEMPLATE.md:16]:
```suggestion

Maintainers: tick if completed or explain if not relevant
```

**Review by Ericson2314 — APPROVED** [review]
